rest/main_new.go

```diff
@@ -0,0 +1,53 @@
+package rest
+
+import (
+	"flag"
+	"fmt"
+
+	"github.com/couchbase/sync_gateway/base"
+	"github.com/imdario/mergo"
+	pkgerrors "github.com/pkg/errors"
+)
+
+// serverMainPersistentConfig runs the Sync Gateway server with persistent config.
+func serverMainPersistentConfig(osArgs []string, fs *flag.FlagSet, flagStartupConfig *StartupConfig) error {
+
+	// 3.0 config - bootstrap and pull configs from server buckets.
+	base.Infof(base.KeyAll, "Running in persistent config mode")
+
+	sc := DefaultStartupConfig(defaultLogFilePath)
+	base.Tracef(base.KeyAll, "default config: %#v", sc)
+
+	fileStartupConfig, err := LoadStartupConfigFromPaths(fs.Args()...)
+	if pkgerrors.Cause(err) == base.ErrUnknownField {
+		// TODO: CBG-1399 Do automatic legacy config upgrade here
+		return fmt.Errorf("Couldn't parse config file: %w (legacy config upgrade not yet implemented)", err)
+	} else if err != nil {
+		return fmt.Errorf("Couldn't open config file: %w", err)
+	}
+
+	if fileStartupConfig != nil {
+		base.Tracef(base.KeyAll, "got config from file: %#v", fileStartupConfig)
+		err := mergo.Merge(&sc, fileStartupConfig, mergo.WithOverride)
+		if err != nil {
+			return err
+		}
+	}
+
+	if flagStartupConfig != nil {
+		base.Tracef(base.KeyAll, "got config from flags: %#v", flagStartupConfig)
+		err := mergo.Merge(&sc, flagStartupConfig, mergo.WithOverride)
+		if err != nil {
+			return err
+		}
+	}
```

Comment:
It's clearer to me what the state of the system is if default values are included in the /_config response, without working backwards to find what the default values are for a given version/edition. Similiarly, it would benefit customers using the /_config endpoint, as it removes the dependency on docs to find what the default values are for a given option.

At least for the file-based config, we should be capturing the raw startup config file to determine exactly what customers are defining via sgcollect.

Reply:
I agree from the perspective of seeing the state of the system, but what's the expected workflow if someone wants to retrieve an existing database config, tweak some properties, and then update?  Assuming they pull down the default values, it looks like they are explicitly setting the defaults in the updated version.  Which has the potential to create problems if we change the default value in a future release.

Reply:
In this case it's the StartupConfig so they can't use the REST API to update, as it's coming from the config file. But I see what you're saying, if users use the REST API to retrieve to write a new config file. I'll have a think about that use-case and see if we're able to cover both bases.